aws-s3-transfer-manager/src/runtime/scheduler.rs

```diff
@@ -3,79 +3,79 @@
+
+// TODO - track high water mark
+// TODO - add statistics/telemetry
 
 /// Manages scheduling networking and I/O work
 ///
 /// Scheduler is internally reference-counted and can be freely cloned.
 #[derive(Debug, Clone)]
 pub(crate) struct Scheduler {
-    sem: Arc<Semaphore>,
+    token_bucket: TokenBucket,
 }
 
 impl Scheduler {
     /// Create a new scheduler with the initial number of work permits.
-    pub(crate) fn new(permits: usize) -> Self {
+    pub(crate) fn new(mode: ConcurrencyMode) -> Self {
         Self {
-            // NOTE: tokio semahpore is fair, permits are given out in the order requested
-            sem: Arc::new(Semaphore::new(permits)),
+            token_bucket: TokenBucket::new(mode),
         }
     }
 
-    // TODO - add some notion of "work type" and/or "work estimate" to permit acquisition to allow
-    // for scheduler to make choices on what work gets prioritized
-
     /// Acquire a permit to perform some unit of work
-    pub(crate) fn acquire_permit(&self) -> AcquirePermitFuture {
-        match self.try_acquire_permit() {
+    pub(crate) fn acquire_permit(&self, ptype: PermitType) -> AcquirePermitFuture {
+        match self.try_acquire_permit(ptype.clone()) {
             Ok(Some(permit)) => AcquirePermitFuture::ready(Ok(permit)),
             Ok(None) => {
                 let inner = self
-                    .sem
-                    .clone()
-                    .acquire_owned()
-                    .map_ok(OwnedWorkPermit::from)
-                    .map_err(error::from_kind(error::ErrorKind::RuntimeError));
-
+                    .token_bucket
+                    .acquire(ptype)
+                    .map_ok(OwnedWorkPermit::from);
                 AcquirePermitFuture::new(inner)
             }
             Err(err) => AcquirePermitFuture::ready(Err(err)),
         }
     }
 
-    /// Try to acquire a permit for some unit of work.
-    ///
-    /// If there are no permits left, this returns `Ok(None)`. Otherwise, this returns
-    /// `Ok(Some(OwnedWorkPermit))`
-    fn try_acquire_permit(&self) -> Result<Option<OwnedWorkPermit>, error::Error> {
-        match self.sem.clone().try_acquire_owned() {
-            Ok(permit) => Ok(Some(permit.into())),
-            Err(err) => match err {
-                TryAcquireError::Closed => {
-                    Err(error::Error::new(error::ErrorKind::RuntimeError, err))
-                }
-                TryAcquireError::NoPermits => Ok(None),
-            },
-        }
+    fn try_acquire_permit(
+        &self,
+        ptype: PermitType,
+    ) -> Result<Option<OwnedWorkPermit>, error::Error> {
+        self.token_bucket
+            .try_acquire(ptype)
+            .map(|token| token.map(OwnedWorkPermit::from))
     }
 }
 
+pub(crate) type PayloadEstimate = u64;
+
+/// The type of work to be done
+#[derive(Debug, Clone)]
+pub(crate) enum PermitType {
+    /// A network request to a control plane API with insignificant payload
+    ControlPlane,
```

Comment:
trivial: maybe don't use these terms? HeadObject is data plane, but has no payload, so ?

Reply:
I think it was suggested to just collapse them to a single variant `Network(Option<PayloadEstimate>)` or similar which covers both use cases.